### ISSUE TITLE
%% anywhere in input is now replaced with %

### ISSUE THREAD

**Issue by kroeschl** [sqlalchemy/mako#384]
Starting with Mako 1.3.1, `%%` anywhere in input now renders to `%`. This is a change in behavior from previous releases.

It's unclear to me if this is intended. [The documentation](https://docs.makotemplates.org/en/latest/syntax.html#control-structures) says that `%%` can be used to render a `%` at the start of a line, but it doesn't explicitly state how `%%` is handled elsewhere in a line. My first interpretation was that `%%` elsewhere in a line wouldn't be changed.

Simple reproducing case:
```
$ pip install mako==1.3.0
...
Successfully installed mako-1.3.0
$ mako-render - <<< foo%%bar
foo%%bar
$ pip install mako==1.3.1
...
Successfully installed mako-1.3.1
$ mako-render - <<< foo%%bar
foo%bar
```

This is related to #383. From the discussion there, it seems like this was expected to possibly break users, but it's not clear to me if it was expected to break all use of `%%` not at the start of a line or just `%%` at the start of a line after whitespace (which I would expect).

**Comment by zzzeek** [comment]
> Starting with Mako 1.3.1, `%%` anywhere in input now renders to `%`. This is a change in behavior from previous releases.

it is.


> 
> It's unclear to me if this is intended. 

it is.

[The documentation](https://docs.makotemplates.org/en/latest/syntax.html#control-structures) says that `%%` can be used to render a `%` at the start of a line, but it doesn't explicitly state how `%%` is handled elsewhere in a line. My first interpretation was that `%%` elsewhere in a line wouldn't be changed.


unfortunately if you look in the example in that documentation, the example itself illustrates two lines with double percents, the second line is indented.     so while the sentence in the paragraph is wrong, the example illustrates what it's supposed to look like.

prior to this change there was simply no way at all to render a single percent sign without enclosing it in quotes.

> > This is related to #383. From the discussion there, it seems like this was expected to possibly break users, but it's not clear to me if it was expected to break all use of `%%` not at the start of a line or just `%%` at the start of a line after whitespace (which I would expect).

it would break for users who were writing exactly two percent signs, or a multiple of two percent signs in a row, and who then expected that exact multiple of percent signs to render.  if you didnt put two percent signs and only put one, or put an odd number of percent signs, you'd get a compile error.  this is clearly buggy behavior and I thought it would be unlikely for a large number of users to have the exact scenario of where they need to render only even numbers of consecutive percent signs as given, who never tried to render an odd number of percent signs.

the only action I can take here, if you are not able to make adjustments to your code, would be that I'd revoke Mako 1.3.1 and re-release as Mako 1.4 with a stronger note.   I dont think the change should be reverted.

**Comment by zzzeek** [comment]
hi -

I just read your case more closely, and I deleted what I wrote since I assumed you were referring to the simple behavioral change of % as the first non-whitespace character.

I will evaluate your case, but I am leaning towards releasing this as Mako 1.4.0 with a stronger note and backwards compatibility note, revoking 1.3.1

**Comment by zzzeek** [comment]
I've yanked the release and reverted the change.

**Comment by kroeschl** [comment]
Perfect, thanks for the quick response!

**Comment by zzzeek** [comment]
please note that the fix **will** still impact double percent signs as the **first non whitespace character** on a line

**Comment by zzzeek** [comment]
a new version of this fix is merged in 1d6c58e7124653bc11b0a4f153bdbaab41622854 and I'm going to put 1.3.2 out with it.  since it's the only change again, if more issues we can keep reverting until we get it right (I haven't spent much time on Mako in over a decade)
